opentelemetry/proto/metrics/v1/metrics.proto

```diff
@@ -238,11 +218,139 @@ message MetricDescriptor {
     //   12. The 1 second collection cycle ends. A metric is exported for the
     //      number of requests received over the interval of time t_1 to
     //      t_0+1 with a value of 1.
-    CUMULATIVE = 3;
+    CUMULATIVE = 2;
+  }
+
+  // Structure is the structural quality of a metric, indicating
+  // whether the input measurements represent part of an individual
+  // sum or are considered individual measurements.
+  enum Structure {
+    // INVALID_STRUCTURE is the default Structure, it MUST not be
+    // used.
+    INVALID_STRUCTURE = 0;
+
+    // GROUPING structure means the value has been computed from
+    // individual input values considered part of a distribution.
+    // GROUPING structure implies the sum of measurements is not
+    // necessarily meaningful.
+    //
+    // Value of this kind are defined as a current measurement of the
+    // metric when represented by single-value data points.
+    GROUPING = 1;
+
+    // ADDING_MONOTONIC structure means the measurement determines a
+    // sum.  For DELTA kind this is expressed as the change in sum
+    // since the last collection.  For CUMULATIVE kind this is
+    // expressed as the last collected value of the sum since
+    // reporting began.
+    //
+    // ADDING_MONOTONIC indicates that the calculated sum is
+    // non-decreasing, therefore can be monitored as a non-negative
+    // rate of change.
+    ADDING_MONOTONIC = 2;
+
+    // ADDING structure means the measurement determines a sum.  For
+    // DELTA kind this is expressed as the change in sum since the
+    // last collection.  For CUMULATIVE kind this is expressed as the
+    // last collected value of the sum since reporting began.
+    ADDING = 3;
+  }
+
+  // Continuity describes how the measurement was captured,
+  // indicating whether measurements represent application-level
+  // events or were generated through a callback invoked by the SDK.
```

Comment:
Is not only the callback that can generate snapshots, aggregating any already aggregated data is a snapshot. Think in collector I can receive points from a Counter aggregated as a sum/delta then building a Histogram there makes the input snapshot

Reply:
Also for a Counter (aggregating cumulative Sum) vs SumObserver (cumulative Sum calculated outside) - both sums were calculated by an aggregator (in our outside our library), and at the moment you export the sum there is minimal information that Snapshot vs Continuous provide - no more information about the number of measurements available. But in case of a Histogram where we also include the count of measurements this information may be useful (and probably is). So I am not 100% convinced that Continuity makes sense for scalar metrics.

Reply:
> aggregating any already aggregated data is a snapshot

I would have stated that:
- aggregating already-aggregated SNAPSHOT data is SNAPSHOT data
- aggregating already-aggregated CONTINUOUS data is CONTINUOUS data

Combining multiple input metrics of different Continuity is not well defined in this (variation of the) protocol, and I'm not sure we should be aiming for a protocol that supports multiple input metrics regardless of whether they agree on Structure/Continuity. In all cases, the presence of StartTimeUnixNano != 0 indicates that the input data spanned a period of time as opposed to being defined at an instant where StartTimeUnixNano == TimeUnixNano.

Reply:
> at the moment you export the sum there is minimal information that Snapshot vs Continuous provide

My intention is that both of these fields tell us about the input measurements, which do not change when we aggregate.

There may be minimal information, but there is information here. If you see two raw measurements with the same timestamp and they are SNAPSHOT, they were made by the same callback and can be compared as a ratio. If you see two raw measurements with the same timestamp and they are CONTINUOUS, they are merely coincidental and should not be compared as a ratio.

A sum is a sum, and we know that it is sensible to compute a rate of a sum. A "gauge" (or non-sum) should not be exposed as a rate, because we don't know that the distance between A and B is the same as the distance between (A+x) and (B+x). This quality is the reason that Sums can be displayed as rates regardless of whether they are monotonic or not.

Reply:
> My intention is that both of these fields tell us about the input measurements, which do not change when we aggregate.

I think this is not clear to me, will try to show an example to show where my confusion comes from:
1. An "UpDownSumObserver" for "memory usage", and having an export interval of 10seconds, every 10 second I report the cumulative sum of all "new/delete" calls.
2. An "UpDownCounter" for "new/delete" calls, by instrumenting tcmalloc library. We have configured a cumulative sum aggregation (maybe without exemplars if that will be possible). Export interval is the same 10 seconds.

In this two cases I get exactly the same data - every 10 second a scalar metric with the sum of new/delete from the beginning of the process. The only difference may be "exemplars" which may or may not be present for the second metric (not guaranteed that they are present, if the user configures to not keep exemplars in the view configuration).

Now the idea is, way do we need to know and complicate the model for this case with `continuous` vs `snapshot`? Both are `snapshots` because even in the second metric we actually take a snapshot of the aggregator at a specific moment of time.


> If you see two raw measurements with the same timestamp and they are SNAPSHOT, they were made by the same callback and can be compared as a ratio. If you see two raw measurements with the same timestamp and they are CONTINUOUS, they are merely coincidental and should not be compared as a ratio.

Here is where I agree with you, but I see things a bit different as proposed in #179, we have "Raw measurements" a.k.a `CONTINUOUS` which represents measurements recorded using the sync instruments with no other aggregation and any other "measurements" are `snapshot` independent of where that aggregation happens (in our library like in the case of the second metric, or outside our library in case of the first metric) because we actually take a snapshot of an `Aggregator`.

Probably I miss something, but here is where my current understanding is.

Reply:
So in my proposal to clarify "Raw Measurements" imply continuos all the other types imply snapshot because we take a snapshot of an Aggregator that somewhere exists (inside or outside of our library)

Reply:
> In this two cases I get exactly the same data - every 10 second a scalar metric with the sum of new/delete from the beginning of the process

The count of label sets means something different in these cases. The Observer instrument produces N points at the same logical instant in time, whereas the Recorder instrument produces N points over a window of time. If I ask how many distinct label sets there are from a Recorder (CONTINUOUS), I have to specify a time window for the question to be meaningful. For an Observer (SNAPSHOT), I can answer the question ("how many distinct label sets") without specifying a time range.

Reply:
> The count of label sets means something different in these cases. The Observer instrument produces N points at the same logical instant in time, whereas the Recorder instrument produces N points over a window of time. If I ask how many distinct label sets there are from a Recorder (CONTINUOUS), I have to specify a time window for the question to be meaningful. For an Observer (SNAPSHOT), I can answer the question ("how many distinct label sets") without specifying a time range.

I think this is where I miss something (maybe here is where the disagreement comes from). In both cases for me when I look at the N points (N different label sets) they accumulated (sums) values over that time interval:

--------  // start time t0'
* t0 - new object with location "stack" size 10
* t1 - new object with location "heap" size 100
* t2 - delete object with location "stack" size 10
* t3 - new object with location "stack" size 20
--------  // exporting t3'
* t4 - delete object with location "stack" size 20
* t5 - new object with location "stack" size 30
* t6 - delete object with location "heap" size 100
* t7 - new object with location "heap" size 200
--------  // exporting t7'
* t8 - delete object with location "stack" size 30
* t9 - delete object with location "heap" size 200
--------  // exporting t9'

First Metric:
* first export (start_time= t0'; end_time = t3' - "location"="stack" 20; "location"="heap" 100)
* second export (start_time= t0'; end_time = t7' - "location"="stack" 30; "location"="heap" 200)
* third export (start_time= t0'; end_time = t9' - "location"="stack" 0; "location"="heap" 0)

Second Metric:
* first export (start_time= t0'; end_time = t3' - "location"="stack" 20; "location"="heap" 100)
* second export (start_time= t0'; end_time = t7' - "location"="stack" 30; "location"="heap" 200)
* third export (start_time= t0'; end_time = t9' - "location"="stack" 0; "location"="heap" 0)

In both cases we have the same number of unique label sets, and we cannot assign a specific moment when a value for a specific label set was updated, because that information is lost during aggregation, the only thing we know is that in the interval t0' - tx' we had this number of distinct label sets and these are the "current values" corresponding to every label set.

To get a bit more confused, even for the "second metric" we actually read the value at one specific moment "exporting time", so why is that different? I think once we left the library by adding the data to the wire, we see these metrics exactly the same, inside the library there may be some interpretations like the one you had, but once we put the data on the wire they are identical to me.

Reply:
I think I misunderstood this:
> In all cases, the presence of StartTimeUnixNano != 0 indicates that the input data spanned a period of time as opposed to being defined at an instant where StartTimeUnixNano == TimeUnixNano.

I am trying again to see if that makes sense:
1. There are some Observers where we know the start time, for example a `SumObserver` may have the start time equal with the start of the process, and in this case their data will look very similar with the "UpDownCounter" example. In this case the metric is considered "CONTINUOUS".
2. There are some Observers where we cannot determine the start time correctly, even for a `SumObserver`, in this case the collected points will miss the start_time and hence they will be considered "SNAPSHOT".

If this is the example that you have in mind, I completely agree that this is a possibility in the real world, but this makes me to think that Observers especially "SumObserver" and "UpDownSumObserver" need to support a configuration where users will specify the start_time (or just a boolean to say if it is equal to the start of the process).